test/self_protocore/upsert_validator.js

```diff
@@ -0,0 +1,151 @@
+// limitations under the License.
+
+'use strict';
+
+const BN = require('bn.js');
+
+const { AccountProvider } = require('../test_lib/utils.js');
+const Utils = require('../test_lib/utils.js');
+
+const TestSelfProtocore = artifacts.require('TestSelfProtocore');
+
+const config = {};
+const validator = {};
+
+contract('SelfProtocore::upsertValidator', (accounts) => {
+  const accountProvider = new AccountProvider(accounts);
+
+  beforeEach(async () => {
+    config.coconsensusAddress = accountProvider.get();
+
+    config.epochLength = new BN(100);
+
+    config.genesisKernelHeight = new BN(1);
+    config.genesisKernelHash = Utils.getRandomHash();
+
+    config.core = accountProvider.get();
+    config.genesis = {};
+    config.genesis.auxiliaryParentVoteMessageHash = Utils.getRandomHash();
+    config.genesis.auxiliarySourceTransitionHash = Utils.getRandomHash();
+    config.genesis.auxiliarySourceBlockHash = Utils.getRandomHash();
+    config.genesis.auxiliaryTargetBlockHash = Utils.getRandomHash();
+    config.genesis.auxiliaryAccumulatedGas = new BN(1000000);
+
+    config.setupParams = {};
+    config.setupParams.metachainId = Utils.getRandomHash();
+    config.setupParams.domainSeparator = Utils.getRandomHash();
+    config.setupParams.epochLength = new BN(100);
+    config.setupParams.metablockHeight = new BN(Utils.getRandomNumber(1000));
+    config.setupParams.selfProtocore = accountProvider.get();
+    config.setupParams.coconsensus = accountProvider.get();
+
+    config.genesis.auxiliarySourceBlockNumber = new BN(
+      Utils.getRandomNumber(10000) * config.setupParams.epochLength,
+    );
+    config.genesis.auxiliaryTargetBlockNumber = config.genesis.auxiliarySourceBlockNumber.add(
+      config.setupParams.epochLength,
+    );
+
+    // Deploy the self protocore contract.
+    config.selfProtocore = await TestSelfProtocore.new();
+
+    await config.selfProtocore.setGenesisStorage(
+      config.genesis.auxiliaryParentVoteMessageHash,
+      config.genesis.auxiliarySourceTransitionHash,
+      config.genesis.auxiliarySourceBlockHash,
+      config.genesis.auxiliarySourceBlockNumber,
+      config.genesis.auxiliaryTargetBlockHash,
+      config.genesis.auxiliaryTargetBlockNumber,
+      config.genesis.auxiliaryAccumulatedGas,
+    );
+    config.openKernelHeight = new BN(20);
+    validator.address = accountProvider.get();
+    validator.beginHeight = config.openKernelHeight.addn(1);
+    validator.endHeight = new BN(100);
+    validator.reputation = new BN(500);
+
+    await config.selfProtocore.setCoconsensus(config.coconsensusAddress);
+    await config.selfProtocore.setOpenKernelHeight(config.openKernelHeight);
+
+    await config.selfProtocore.setup(
+      config.setupParams.metachainId,
+      Utils.getRandomHash(),
+      new BN(100),
+      new BN(20),
+      {
+        from: config.coconsensusAddress,
+      },
+    );
+  });
+
+  contract('Positive Tests', async () => {
```

Comment:
there are definitely more unit tests valuable here for the different edge-cases of upsert, covering combinations of ({Height1, Height2}, Validator, {RepZero, Rep1, Rep2}).

@jayeshbairagi can you note this as a new ticket ? (either refer to this comment, or put it in more sensible wording :D )

Reply:
Done, noted for further discussion.